File: bounty-program/stellar-docs/python/clawback.py
from stellar_sdk import (
    Server,
    Keypair,
    TransactionBuilder,
    Network,
    Asset,
    AuthorizationFlag,
    SetOptions,
    ChangeTrust,
    Payment,
    Clawback
)

server = Server("https://horizon-testnet.stellar.org")

A = Keypair.from_secret("SARQWRSMT5REPFQHNBHSPX7WICDJZZ43GD4YICYBNKIA7WZB3MTHAPZL")
B = Keypair.from_secret("SDFKWXO3WRD5ZHS4G5IZYFOBBUYPMUKFNMK5TTKGIMSGHAZ4CPDGJMM7")
C = Keypair.from_secret("SB7QOMOLKBXGOKE5JUK26ZO5ANPAF7OLE76ZT5Y7MSCK6CYQNVRCQIBR")

ASSET = Asset(code = "CLAW", issuer = A.public_key)

# Enables AuthClawbackEnabledFlag on an account.
def enableClawback(account, keys):
    tx = buildTx(
            source = account,
            signer = keys,
            ops = [
                SetOptions(
                    set_flags = AuthorizationFlag.AUTHORIZATION_CLAWBACK_ENABLED |
                        AuthorizationFlag.AUTHORIZATION_REVOCABLE
                )
            ]
    )
    return server.submit_transaction(tx)

# Establishes a trustline for `recipient` for ASSET (from above).
def establishTrustline(recipient, key):
    tx = buildTx(
            source = recipient,
            signer = key,
            ops = [
                ChangeTrust(
                    asset = ASSET,
                    limit = "5000" # arbitrary
                )
            ]
    )
    return server.submit_transaction(tx)

# Retrieves latest account info for all accounts.
def getAccounts():
    return(
        server.load_account(A.public_key),
        server.load_account(B.public_key),
        server.load_account(C.public_key)
    )

# Enables clawback on A, and establishes trustlines from C, B -> A.
def preamble():
    accountA, accountB, accountC = getAccounts()
    return(
        enableClawback(accountA, A),
        establishTrustline(accountB, B),
        establishTrustline(accountC, C)
    )

# Helps simplify creating & signing a transaction.
def buildTx(source, signer, ops):
    tx = (
        TransactionBuilder (
            source_account = source,
            network_passphrase = Network.TESTNET_NETWORK_PASSPHRASE,
            base_fee = server.fetch_base_fee()
        )
    )
    for op in ops:
        tx.append_operation(op)
    tx = tx.set_timeout(30).build()
    tx.sign(signer)
    return tx

# Prints the balances of a list of accounts.
def showBalances(accounts):
    for acc in accounts:
        print(acc.account_id()[0:5] + ": " + getBalance(acc))

# Make a payment to `toAccount` from `fromAccount` for `amount`.
def makePayment(toAccount, fromAccount, fromKey, amount):
    tx = buildTx(
            source = fromAccount,
            signer = fromKey,
            ops = [
                Payment(
                    destination = toAccount.account_id(),
                    asset = ASSET,
                    amount = amount
                )
            ]
    )
    return server.submit_transaction(tx)

# Perform a clawback by `byAccount` of `amount` from `fromAccount`.
def doClawback(byAccount, byKey, fromAccount, amount):
    tx = buildTx(
            source = byAccount,
            signer = byKey,
            ops = [
                Clawback(
                    asset = ASSET,
                    from_ = fromAccount.account_id(),
                    amount = amount
                )
            ]
    )
    return server.submit_transaction(tx)

    # Retrieves the balance of ASSET in `account`.
def getBalance(account):
    balances = server \
        .accounts() \
        .account_id(account.account_id()) \
        .call()["balances"]

    for i in range(len(balances)):
        if balances[i].get("asset_code") == ASSET.code and balances[i].get("asset_issuer") == ASSET.issuer:
            return balances[i]["balance"]
    return "0"

def examplePaymentClawback():
    accountA, accountB, accountC = getAccounts()
    return(
        makePayment(accountB, accountA, A, "1000"),
        makePayment(accountC, accountB, B, "500"),
        doClawback(accountA, A, accountC, "250")
    )

if __name__ == "__main__":
    preamble()
    examplePaymentClawback()
    showBalances(getAccounts())

File: bounty-program/stellar-docs/java/clawback.java
import org.stellar.sdk.*;
import org.stellar.sdk.responses.AccountResponse;
import org.stellar.sdk.responses.SubmitTransactionResponse;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Clawback {

    static Server server = new Server("https://horizon-testnet.stellar.org");

    static KeyPair A = KeyPair.fromSecretSeed("SCF3XYHHYI6CMVXMKU7AQKUU5F7F6MFEC6CXKM4C44YKYLFKBZRLMBME");
    static KeyPair B = KeyPair.fromSecretSeed("SA5LIXGGDSSMT3QJU6GEYV2XMN5U3QTJNRNQAAIX5M7TOZN5A6KDR5BJ");
    static KeyPair C = KeyPair.fromSecretSeed("SDFIA4SZTPJNEMYLXRZTIWO3T5IG7SF43SMKCVEBRE7S2BKW6FFKJTMX");

    static Asset ASSET = Asset.createNonNativeAsset("CLAW", A.getAccountId());

    // Enables AuthClawbackEnabledFlag on an account.
    static SubmitTransactionResponse enableClawback(TransactionBuilderAccount account, KeyPair keys) {
        try {
            List<Operation> operationList = new ArrayList<>();
            operationList.add(
                    new SetOptionsOperation.Builder()
                            .setSetFlags(2 | 8)
                            .build()
            );
            Transaction tx = buildTx(account, keys, operationList);
            return server.submitTransaction(tx);
        }
        catch (IOException | AccountRequiresMemoException e) {
            e.printStackTrace();
            return null;
        }
    }

    /// Establishes a trustline for `recipient` for ASSET (from above).
    static SubmitTransactionResponse establishTrustline(TransactionBuilderAccount recipient, KeyPair key) {

        try {
            List<Operation> operationList = new ArrayList<>();
            operationList.add(
                    new ChangeTrustOperation.Builder(ASSET, "5000").build()
            );
            Transaction tx = buildTx(recipient, key, operationList);
            return server.submitTransaction(tx);
        }
        catch (IOException | AccountRequiresMemoException e) {
            e.printStackTrace();
            return null;
        }
    }

    /// Retrieves the latest account info for all accounts.
    static TransactionBuilderAccount[] getAccounts() {
        try {
            return new TransactionBuilderAccount[]{
                    server.accounts().account(A.getAccountId()),
                    server.accounts().account(B.getAccountId()),
                    server.accounts().account(C.getAccountId())
            };
        }
        catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    /// Enables clawback on A, and establishes trustlines from C, B -> A.
    static void preamble() {

        TransactionBuilderAccount accountA = getAccounts()[0];
        TransactionBuilderAccount accountB = getAccounts()[1];
        TransactionBuilderAccount accountC = getAccounts()[2];

        enableClawback(accountA, A);
        establishTrustline(accountB, B);
        establishTrustline(accountC, C);
    }

    /// Helps simplify creating & signing a transaction.
    static Transaction buildTx(TransactionBuilderAccount source, KeyPair signer, List<Operation> ops) {
        Transaction.Builder txBuilder = new Transaction.Builder(source, Network.TESTNET)
                .setBaseFee(Transaction.MIN_BASE_FEE);
        for (Operation op : ops) {
            txBuilder.addOperation(op);
        }
        Transaction tx = txBuilder.setTimeout(30).build();
        tx.sign(signer);
        return tx;
    }

    /// Prints the balances of a list of accounts.
    static void showBalances(TransactionBuilderAccount[] accounts) {
        for(TransactionBuilderAccount account: accounts)
            System.out.println(account.getAccountId().substring(0, 5) + ": "
                    + Float.parseFloat(getBalance(account)));
    }

    /// Make a payment to `toAccount` from `fromAccount` for `amount`.
    static SubmitTransactionResponse makePayment(TransactionBuilderAccount toAccount,
                                                 TransactionBuilderAccount fromAccount, KeyPair fromKey,
                                                 String amount) {
        try {
            List<Operation> operationList = new ArrayList<>();
            operationList.add(
                    new PaymentOperation.Builder(
                            toAccount.getAccountId(), ASSET, amount).build()
            );
            Transaction tx = buildTx(fromAccount, fromKey, operationList);
            return server.submitTransaction(tx);
        }
        catch (IOException | AccountRequiresMemoException e) {
            e.printStackTrace();
            return null;
        }
    }

    /// Perform a clawback by `byAccount` of `amount` from `fromAccount`.
    static SubmitTransactionResponse doClawback(TransactionBuilderAccount byAccount, KeyPair byKey,
                                                TransactionBuilderAccount fromAccount, String amount) {
        try {
            List<Operation> operationList = new ArrayList<>();
            operationList.add(
                    new ClawbackOperation.Builder(
                            fromAccount.getAccountId(), ASSET, amount).build()
            );
            Transaction tx = buildTx(byAccount, byKey, operationList);
            return server.submitTransaction(tx);
        }
        catch (IOException | AccountRequiresMemoException e) {
            e.printStackTrace();
            return null;
        }
    }

    /// Retrieves the balance of ASSET in `account`.
    static String getBalance(TransactionBuilderAccount account) {
        try {
            AccountResponse.Balance[] balances = server.accounts()
                    .account(account.getAccountId()).getBalances();

            for(AccountResponse.Balance balance: balances) {
                if(Objects.equals(balance.getAssetCode(), "CLAW")
                        && Objects.equals(balance.getAssetIssuer(), A.getAccountId())) {
                    return balance.getBalance();
                }
            }
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        return "0";
    }

    static void examplePaymentClawback() {
        TransactionBuilderAccount accountA = getAccounts()[0];
        TransactionBuilderAccount accountB = getAccounts()[1];
        TransactionBuilderAccount accountC = getAccounts()[2];

        makePayment(accountB, accountA, A, "1000");
        makePayment(accountC, accountB, B, "500");
        doClawback( accountA, A, accountC, "250");

        showBalances(getAccounts());
    }

    public static void main(String[] args) {
        preamble();
        examplePaymentClawback();
    }
}
